File: FWFramework/Classes/Component/AppleMusic/FWAuthorizeAppleMusic.m
//
//  FWAuthorize.m
//  FWFramework
//
//  Created by wuyong on 17/3/15.
//  Copyright © 2018年 wuyong.site. All rights reserved.
//

#import "FWAuthorizeAppleMusic.h"
#import <MediaPlayer/MediaPlayer.h>

// iOS9.3+需要授权，9.3以前不需要授权
@interface FWAuthorizeAppleMusic : NSObject <FWAuthorizeProtocol>

@end

@implementation FWAuthorizeAppleMusic

+ (void)load
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        [FWAuthorizeManager registerAuthorize:FWAuthorizeTypeAppleMusic withBlock:^id<FWAuthorizeProtocol>{
            return [[FWAuthorizeAppleMusic alloc] init];
        }];
    });
}

- (FWAuthorizeStatus)authorizeStatus
{
    MPMediaLibraryAuthorizationStatus status = [MPMediaLibrary authorizationStatus];
    switch (status) {
        case MPMediaLibraryAuthorizationStatusRestricted:
            return FWAuthorizeStatusRestricted;
        case MPMediaLibraryAuthorizationStatusDenied:
            return FWAuthorizeStatusDenied;
        case MPMediaLibraryAuthorizationStatusAuthorized:
            return FWAuthorizeStatusAuthorized;
        case MPMediaLibraryAuthorizationStatusNotDetermined:
        default:
            return FWAuthorizeStatusNotDetermined;
    }
}

- (void)authorize:(void (^)(FWAuthorizeStatus status))completion
{
    [MPMediaLibrary requestAuthorization:^(MPMediaLibraryAuthorizationStatus status) {
        if (completion) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completion(self.authorizeStatus);
            });
        }
    }];
}

@end

File: FWFramework/Classes/Module/AppleMusic/FWAuthorizeAppleMusic.swift
//
//  FWAuthorizeAppleMusic.swift
//  FWFramework
//
//  Created by wuyong on 2019/6/27.
//  Copyright © 2019 wuyong.site. All rights reserved.
//

import MediaPlayer
#if FWFrameworkSPM
import FWFramework
import FWFrameworkCompatible
#endif

/// AppleMusic授权
@objcMembers public class FWAuthorizeAppleMusic: NSObject, FWAuthorizeProtocol, FWAutoloadProtocol {
    public static func autoload() {
        FWAuthorizeManager.registerAuthorize(.appleMusic) {
            return FWAuthorizeAppleMusic()
        }
    }
    
    public func authorizeStatus() -> FWAuthorizeStatus {
        let status = MPMediaLibrary.authorizationStatus()
        switch status {
        case .denied:
            return .denied
        case .restricted:
            return .restricted
        case .authorized:
            return .authorized
        default:
            return .notDetermined
        }
    }
    
    public func authorize(_ completion: ((FWAuthorizeStatus) -> Void)?) {
        MPMediaLibrary.requestAuthorization { status in
            if completion != nil {
                DispatchQueue.main.async { [weak self] in
                    guard let self = self else { return }
                    completion?(self.authorizeStatus())
                }
            }
        }
    }
}
